File: bindings/csharp/src/Sone/Dsl.cs
namespace Sone;

/// <summary>
/// The node factories. Add <c>using static Sone.Dsl;</c> and the tree reads the
/// way it does in TypeScript:
/// <code>
/// var root = Column(
///     Text("Hello").Size(28).Weight(FontWeight.Bold),
///     Row(
///         Column().Bg("salmon").Size(50).Rounded(14),
///         Column().Bg("orange").Size(50).Rounded(14)
///     ).Gap(10)
/// ).Gap(20).Padding(20).Bg("khaki").CornerRadius(28);
/// </code>
/// </summary>
/// <remarks>
/// The factory and the class it returns are named separately —
/// <c>Column()</c> returns a <see cref="ColumnNode"/> — because a static method
/// and a type cannot share a name and still be invocable. It is the same split
/// the Python binding uses, and it happens to defuse most BCL collisions too;
/// the one that survives is <c>Path</c>, which is <see cref="System.IO.Path"/>,
/// so the factory here is <see cref="SvgPath"/>.
/// </remarks>
public static class Dsl
{
    /// <summary>A vertical container. Null children are dropped, so <c>cond ? Foo() : null</c> works.</summary>
    public static ColumnNode Column(params INode?[] children) => Build(new ColumnNode(), children);

    /// <summary>A horizontal container.</summary>
    public static RowNode Row(params INode?[] children) => Build(new RowNode(), children);

    /// <summary>A grid container with row-major auto placement.</summary>
    public static GridNode Grid(params INode?[] children) => Build(new GridNode(), children);

    /// <summary>A paragraph of strings and <see cref="SpanNode"/>s.</summary>
    public static TextNode Text(params Inline[] content)
    {
        var node = new TextNode();
        foreach (var inline in content)
        {
            node.InlineContent.Add(inline);
        }
        return node;
    }

    /// <summary>A styled run inside a <see cref="TextNode"/>.</summary>
    public static SpanNode Span(string text)
    {
        var node = new SpanNode();
        node.InlineContent.Add(text);
        return node;
    }

    /// <summary>Cascade text styling onto every descendant, drawing no box.</summary>
    public static TextDefaultNode TextDefault(params INode?[] children) =>
        Build(new TextDefaultNode(), children);

    /// <summary>An image, from a path, a URL, or <c>asset:name</c>.</summary>
    public static PhotoNode Photo(string src)
    {
        var node = new PhotoNode();
        node.Props.Set("src", src);
        return node;
    }

    /// <summary>An image from raw bytes, inlined into the document as a data URL.</summary>
    public static PhotoNode Photo(ReadOnlySpan<byte> data) =>
        Photo("data:application/octet-stream;base64," + Convert.ToBase64String(data));

    /// <summary>An SVG path. Named for <see cref="System.IO.Path"/>, which owns <c>Path</c>.</summary>
    public static SvgPathNode SvgPath(string d)
    {
        var node = new SvgPathNode();
        node.Props.Set("d", d);
        return node;
    }

    /// <summary>A table. Children are rows.</summary>
    public static TableNode Table(params INode?[] rows) => Build(new TableNode(), rows);

    /// <summary>A table row. Children are cells.</summary>
    public static TableRowNode TableRow(params INode?[] cells) => Build(new TableRowNode(), cells);

    /// <summary>A table cell.</summary>
    public static TableCellNode TableCell(params INode?[] children) => Build(new TableCellNode(), children);

    /// <summary>A bulleted or numbered list. Children are items.</summary>
    public static ListNode List(params INode?[] items) => Build(new ListNode(), items);

    /// <summary>One item in a list.</summary>
    public static ListItemNode ListItem(params INode?[] children) => Build(new ListItemNode(), children);

    /// <summary>Clip every child to an SVG path.</summary>
    public static ClipGroupNode ClipGroup(string path, params INode?[] children)
    {
        var node = Build(new ClipGroupNode(), children);
        node.Props.Set("clipPath", path);
        return node;
    }

    /// <summary>An explicit page break. Only meaningful with <c>pageHeight</c> set.</summary>
    public static ColumnNode PageBreak() =>
        Column().Height(0).PageBreak(PageBreakMode.Before);

    private static T Build<T>(T node, INode?[] children) where T : Node
    {
        node.Adopt(children);
        return node;
    }
}

File: bindings/ruby/lib/sone/dsl.rb
# frozen_string_literal: true

require_relative "nodes"

module Sone
  # Child factories, available inside any container's block.
  #
  # A factory appends a child and returns it, so `column { ... }` reads as part
  # of the parent's block rather than as an assignment.
  module Children
    CONTAINERS = {
      column: ColumnNode,
      row: RowNode,
      grid: GridNode,
      text_default: TextDefaultNode,
      table: TableNode,
      table_row: TableRowNode,
      table_cell: TableCellNode,
      list: ListNode,
      list_item: ListItemNode
    }.freeze

    CONTAINERS.each do |name, klass|
      define_method(name) { |&block| adopt(klass.new, &block) }
    end

    # A paragraph. Positional strings become its content.
    def text(*content, &block)
      node = TextNode.new
      content.each { |item| node.inline << item.to_s }
      adopt(node, &block)
    end

    def photo(src, &block)
      adopt(PhotoNode.new.src(src.to_s), &block)
    end

    def path(d, &block)
      adopt(PathNode.new.d(d.to_s), &block)
    end
    alias svg_path path

    def clip_group(clip_path, &block)
      adopt(ClipGroupNode.new.clip_path(clip_path.to_s), &block)
    end

    # An explicit page break. Named with a bang because `page_break` is already
    # the property that decides where breaks may fall.
    def page_break!
      adopt(ColumnNode.new.height(0).page_break(:before))
    end

    # Append an already-built node — the hook for helper methods that return a
    # subtree.
    def append(child)
      children << child
      child
    end
    alias << append

    private

    def adopt(child, &block)
      child.configure(&block)
      children << child
      child
    end
  end

  # Paragraph content, for `text` and `span`.
  module InlineContent
    # Append raw text.
    def content(*strings)
      strings.each { |string| inline << string.to_s }
      self
    end

    # Append a styled run.
    def span(text = nil, &block)
      node = SpanNode.new
      node.inline << text.to_s unless text.nil?
      node.configure(&block)
      inline << node
      node
    end
  end

  [ColumnNode, RowNode, GridNode, TextDefaultNode, TableNode, TableRowNode,
   TableCellNode, ListNode, ListItemNode, ClipGroupNode].each do |klass|
    klass.include(Children)
  end

  [TextNode, SpanNode].each { |klass| klass.include(InlineContent) }

  # ── top-level factories ───────────────────────────────────────────────────
  #
  # `Sone.column do ... end` builds a root. Nothing is defined on Object, so
  # there is no import ceremony and nothing to collide with.

  Children::CONTAINERS.each do |name, klass|
    define_singleton_method(name) { |&block| klass.new.configure(&block) }
  end

  def self.text(*content, &block)
    node = TextNode.new
    content.each { |item| node.inline << item.to_s }
    node.configure(&block)
  end

  def self.span(text = nil, &block)
    node = SpanNode.new
    node.inline << text.to_s unless text.nil?
    node.configure(&block)
  end

  def self.photo(src, &block)
    PhotoNode.new.src(src.to_s).configure(&block)
  end

  def self.path(d, &block)
    PathNode.new.d(d.to_s).configure(&block)
  end

  class << self
    alias svg_path path
  end

  def self.clip_group(clip_path, &block)
    ClipGroupNode.new.clip_path(clip_path.to_s).configure(&block)
  end

  def self.page_break
    ColumnNode.new.height(0).page_break(:before)
  end
end
